### PR TITLE
Build hp df with attributes

### PR DESCRIPTION
**Pull request by tyluRp**
This carries the data forward by preserving an attribute `source` which represent the original dataset passed to `add_root()`. I also removed some dplyr functions to avoid depending on dplyr. Also added the pipe as a dependency since we have functions that are designed to use them. Now we can do:

```r
devtools::load_all()

df <- structure(
  list(
    Survey = c(
      "OS Students 2014/15",
      "OS Students 2014/15",
      "OS Students 2014/15",
      "OS Students 2014/15",
      "OS Students 2014/15",
      "OS Students 2014/15",
      "OS Students 2014/15"
    ),
    `Operating System` = c("OS X",
                           "OS X", "Linux", "Linux", "Windows", "Windows", "Windows"),
    `OS Version` = c(
      "Yosemite",
      "Leopard",
      "Debian",
      "Ubuntu",
      "Win7",
      "Win8",
      "Win10"
    ),
    users = c(16,
              43, 27, 36, 31, 32, 4)
  ),
  row.names = c(NA, -7L),
  class = c("data.frame")
)

df %>% 
  add_root("OS Students 2014/15") %>% 
  add_layer(
    parent_vals = "OS Students 2014/15",
    child_col = "Operating System",
    link_vals = "OS", 
    node_type_vals = "OS"
  ) %>% 
  add_layer(
    parent_col = "Operating System",
    child_col = "OS Version",
    link_vals = "Ver",
    node_type_vals = "Sub",
    attribute_cols = "users"
  ) %>% 
  hp_dataframe(
    title = "Survey Results of Most Popular OS in 2014/15",
    settings = hierplane_settings(attributes = "attribute1"),
    styles = hierplane_styles(
      link_to_positions = list(Ver = "right")
    )
  ) %>%
  hierplane()
```

### CODE REVIEW

**Comment by tyluRp** [comment]
As I look at this example, I am a little confused about using `parent_vals` in the first layer and then `parent_cols` in the second. How do we explain this?

**Comment by mathidachuk** [comment]
The "Survey" column does not always exist in datasets. So let's say if it does not exist, we will need to manually define the parent value to link the top level nodes to. 

In this case tho, the parent can be defined as `parent_vals = "OS Students 2014/15"` or `parent-col = "Survey"`

**Comment by tyluRp** [comment]
Oh so the first layer could just use parent_col="Survey"? I prefer that much more because it's consistent, and you can just reference the child_col from the above layer when defining the parent_col in the next layer.

**Comment by mathidachuk** [comment]
Yup that's the intent! But data doesn't always come in the format we need to so I built in some flexibility :)

File: R/builder.R [hunk 1]
```diff
@@ -94,10 +92,10 @@ add_layer <- function(.data,
     }
   }
 
-  out %>%
-    filter(!is.na(child),
-           !is.na(link),
-           !is.na(node_type))
+  layer <- out[!is.na(out$child) & !is.na(out$link) & !is.na(out$node_type), ]
```
Comment (mathidachuk) [inline R/builder.R:95]: lol....im seeing my weird dependency on dplyr once you rewrite everything in base r.

**Pull request merged**